src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -204,31 +195,34 @@
  * stable function will act like a cache for the original {@linkplain Function}:
  *
  * {@snippet lang = java:
- * public final class SqrtUtil {
+ * class Log2Util {
  *
- *     private SqrtUtil() {}
+ *     private Log2Util() {}
  *
- *     private static final Set<Integer> CACHED_KEYS = Set.of(1, 2, 4, 8, 16, 32);
+ *     private static final Set<Integer> CACHED_KEYS =
+ *             Set.of(1, 2, 4, 8, 16, 32);
+ *     private static final UnaryOperator<Integer> LOG2_ORIGINAL =
+ *             i -> 31 - Integer.numberOfLeadingZeros(i);
  *
- *     private static final Function<Integer, Double> SQRT =
+ *     private static final Function<Integer, Integer> LOG2_CACHED =
  *             // @link substring="function" target="#function(Set,Function)" :
- *             StableValue.function(CACHED_KEYS, StrictMath::sqrt);
+ *             StableValue.function(CACHED_KEYS, LOG2_ORIGINAL);
  *
- *     public static double sqrt(int a) {
- *          if (CACHED_KEYS.contains(a)) {
- *              return SQRT.apply(a);
- *          } else {
- *              return StrictMath.sqrt(a);
- *          }
+ *     public static double log2(int a) {
```

Comment:
IMHO this example reveals a lack of compositionality of the partial functions we have. If the functions returned optional (to indicate "not defined"), then code like this would become much simpler to express:

```
return LOG2_CACHED.apply(a)
                                     .orElse(LOG2_ORIGINAL.apply(a));
```

It's not just that this is more fluent -- this is also using the caching function to check whether the input is cached or not -- instead of having a separate check (potentially duplicating the costs).

Reply:
I agree with this observation. We should look at providing true partial functions using for example `Optional`.

Reply:
Btw, I believe for now we should just "note" the issue, and move one. Let's have the javadoc reflect what the impl really does. We have good ideas on how to address this, so we can either follow up with a CSR post-integration, or we can address this in the next round of preview

Reply:
(That said, perhaps the javadoc would be better off in showing examples that did not rely on totalizing partial behavior. The examples in the JEP don't do that, and so doesn't fibonacci. We should find a couple more like those).